keps/sig-api-machinery/20190724-http-status-check.md

```diff
@@ -0,0 +1,204 @@
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Current HTTP probe is not sufficient for check HTTP service running correct or not. In some
+situation user need to add accuracy status code and the HTTP response content to judge
+the web service is work well or not.
+
+HTTP Probe is a mechanism of probe pod is running correct or not. It is part of probe.
+Usually HTTP response is between 200 and 400 is correct
+
+This KEP is proposing add some extra info to check HTTP service was work correct or not.
+
+## Motivation
+Sometimes people need a accurate status code check for the HTTP probe. Such as I run the
+ElasticSearch on the kubernetes. When there is a special request, which has a large result.
+It can trigger the the circuit breaker. It cause all request to the ElasticSearch return
+429 temporary.
+
+Another is some special program when it encounter a problem it will return 302 instead of 
+another code.
+
+There is also have some web program has it's owner healthy check page. It return success
+or false in the HTTP response content. So we need a mechanism to check the HTTP probe 
+response is same as I expect or not.
```

Comment:
Rather than extending the kubernetes API to cover such cases, could that be adapted with an exec probe check that would make the request and detect the relevant text in the response content? In general, alternatives like that should be considered before adding to the API surface area. If that is insufficient, that would be good thing to include in an "alternatives considered" section of this proposal along with the reasons why.

Reply:
Yes, use exec probe check will solve this cause.   But use exec probe check also can instead of HTTP probe check. User can add the script to detect the web service is available or not. Why kubernete add a HTTP probe?  I think the reason is HTTP use every where, add HTTP probe will make user easy to use the kubernetes. Now the HTTP probe seems was little deficiencies. We should consider fix this problem. That was my opinion.